### PR TITLE
Implementation of the StableDag data structure

### PR DESCRIPTION
**Pull request by davenza**
This is an early implementation of a StableDag.The code is almost copied from the Dag structure, but using StableGraph instead of Graph.  As indicated in the original commit, there are 4 methods that are not implemented because StableGraph does not provide the necessary functions:

    raw_nodes() -> Because StableGraph lacks raw_nodes().
    node_weights_mut() -> Because StableGraph lacks node_weights_mut().
    raw_edges() -> Because StableGraph lacks raw_edges().
    edge_weights_mut() -> Because StableGraph lacks edge_weights_mut().

I included the same set of tests for StableDag that is used in the Dag structure in the files
tests/add_edges_stable.rs
tests/iterators_stable.rs
tests/walkers_stable.rs

I also included the tests in tests/stable_indices to test that node/edge indices do not change when we remove an edge/node.

I am pretty new to Rust, so it may need some better code / project organization.

### CODE REVIEW

File: src/stabledag.rs [hunk 1]
```diff
@@ -0,0 +1,936 @@
...
+    pub fn edge_weight(&self, edge: EdgeIndex<Ix>) -> Option<&E> {
+        self.graph.edge_weight(edge)
+    }
+
+    /// Mutably borrow the weight from the edge at the given index.
+    pub fn edge_weight_mut(&mut self, edge: EdgeIndex<Ix>) -> Option<&mut E> {
+        self.graph.edge_weight_mut(edge)
+    }
+
+    /// Read from the internal edge array.
```
Comment (mitchmindtree) [inline src/stabledag.rs]: We probably don't need these commented methods - we likely won't ever implement them as I believe the way edges are stored differs between the `Graph` and `StableGraph`.

File: src/stabledag.rs [hunk 2]
```diff
@@ -0,0 +1,936 @@
...
+    _phantom: PhantomData<Ix>,
+}
+
+/// An alias to simplify the **Recursive** **Walker** type returned by **StableDag**.
+pub type RecursiveWalk<N, E, Ix, F> = walker::Recursive<StableDag<N, E, Ix>, F>;
+
+/// An error returned by the `StableDag::add_edge` method in the case that adding an edge would have
+/// caused the graph to cycle.
+#[derive(Copy, Clone)]
+pub struct WouldCycle<E>(pub E);
```
Comment (mitchmindtree) [inline src/stabledag.rs]: I think I'd prefer if we re-used the `WouldCycle` error from the crate root rather than declaring a new one for `StableDag`. It might just make error handling downstream a bit easier in cases where the user is working with both `Dag`s and `StableDag`s.

File: src/stabledag.rs [hunk 3]
```diff
@@ -0,0 +1,936 @@
...
+    type EdgesDirected = Edges<'a, E, Ix>;
+    fn edges_directed(self, a: Self::NodeId, dir: pg::Direction) -> Self::EdgesDirected {
+        self.graph.edges_directed(a, dir)
+    }
+}
+
+impl<'a, N, E, Ix> IntoNodeIdentifiers for &'a StableDag<N, E, Ix>
+    where
+        Ix: IndexType
+
```
Comment (mitchmindtree) [inline src/stabledag.rs]: unnecessary space

**Review by mitchmindtree — CHANGES_REQUESTED** [review]
Thanks so much @davenza, this is really great stuff! I've been meaning to get around to this for a while, I can definitely see myself using this down the track.

I just have a couple very minor things to clean up - otherwise I'm more than happy to merge!

So sorry for the delay by the way, it's getting a little overwhelming maintaining what must be ~50 crates by now, whilst also working at my own business! Thanks for being patient with me :)

**Comment by davenza** [comment]
Hi, thanks for you review. I made the changes you asked for. They are in 5482078. 

Then, I added a feature called `stable_dag` for conditional compilation in cf752f4. What do you think about this?

**Comment by davenza** [comment]
I forgot to include conditional compilation in the tests, but it should be solved in cce14bc.

**Review by davenza — COMMENTED** [review]
Reviewed changes.

**Comment by fwcd** [comment]
Any updates on this PR? I would love to help if there are any specific issues, especially in regard to https://github.com/RustAudio/dsp-chain/issues/103.

**Comment by mitchmindtree** [comment]
Thanks for the reminder @fwcd!

**Review by mitchmindtree — APPROVED** [review]
